### PR TITLE
XNNPACK - GEMM & MATMUL integration 

### PR DESCRIPTION
**Pull request by shalvamist**
### Description - DRAFT
Added support for XNNPACK GEMM & MATMUL ops.
Tested locally on mobileBert & wasm unitests. (right now there is an issue with WASM unitests) looking into this right now.

### Motivation and Context
We documented a ~7% performance improvement on mobileBert using XNNpack

### CODE REVIEW

File: js/web/package-lock.json [hunk 1]
```diff
@@ -91,9 +91,9 @@
       }
     },
     "node_modules/@electron/get": {
-      "version": "1.13.1",
```
Comment (fs-eire) [inline js/web/package-lock.json:94]: It probably is not a bad idea to upgrade the package version, but should not in this PR.

File: onnxruntime/core/providers/cpu/math/gemm.cc [hunk 2]
```diff
@@ -169,6 +169,7 @@ Status Gemm<float>::PrePack(const Tensor& tensor, int input_idx,
       prepacked_weights->buffer_sizes_.push_back(packed_b_size);
     }
   }
+
```
Comment (fs-eire) [inline onnxruntime/core/providers/cpu/math/gemm.cc:172]: looks like it's safe to revert every change in this file. (blank lines and `#if 0`s)

File: onnxruntime/core/providers/xnnpack/detail/node_support_checker.cc [hunk 3]
```diff
@@ -17,6 +17,8 @@
 // each operator provides a helper to check if supported
 #include "core/providers/xnnpack/nn/conv.h"
 #include "core/providers/xnnpack/nn/max_pool.h"
+#include "core/providers/xnnpack/nn/gemm.h"
```
Comment (fs-eire) [inline onnxruntime/core/providers/xnnpack/detail/node_support_checker.cc:20]: gemm and matmul should be under path `xnnpack/math/` instead of `xnnpack/nn`. see other folders like `cpu` or `cuda` under `providers` folder.

File: onnxruntime/core/providers/xnnpack/nn/gemm.cc [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+namespace onnxruntime {
+namespace xnnpack {
+
+bool Gemm::IsGemmOnnxNodeSupported(const NodeUnit& node_unit, const GraphViewer& graph) {
+
+  bool supported = false;
+  const onnxruntime::Node& node = node_unit.GetNode();
+
+  // use do {} while(false) so it's easier to set a breakpoint on the return
```
Comment (fs-eire) [inline onnxruntime/core/providers/xnnpack/nn/gemm.cc:22]: why not just `return true` in every place for `break`?

File: onnxruntime/core/providers/xnnpack/nn/gemm.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+  Status Compute(OpKernelContext* /*context*/) const override;
+
+  static bool IsGemmOnnxNodeSupported(const NodeUnit& node_unit, const GraphViewer& graph);
+
+  Status PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
+                 /*out*/ bool& is_packed,
+                 /*out*/ PrePackedWeights* prepacked_weights) override;
+
+  Status CreateXnnpackOpp(
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/nn/gemm.h:31]: nit: CreateXnnpackOp. 

Is this required though? Can't see any usage of it.

File: onnxruntime/core/providers/xnnpack/nn/gemm.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      const float* kernel,
+      const float* bias,
+      float output_min,
+      float output_max,
+      uint32_t flags);
+
+ private:
+
+  TensorShape b_shape_;
+  BufferUniquePtr packed_b_=nullptr;
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/nn/gemm.h:45]: nit: please format with clang-format. Also a unique_ptr doesn't need explicit initialization.

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+    // B matrix must be constant
+    if (B_arg.Exists() && graph.GetConstantInitializer(B_arg.Name(), true) == nullptr) {
+        break;
+    }
+
+    // making sure we are dealing with MatMul
+    const auto* B_shape = B_arg.Shape();
+
+    if (!B_shape || B_shape->dim_size() > 3) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/nn/matmul.cc:51]: Can you explain this check? I thought the MatMulComputeHelper handled a range of different ranks and limiting this to 2 or 3D wouldn't be necessary. Naively if anything I would expect it to be limited to 2D though.

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    const auto* B_type = B_arg.TypeAsProto();
+
+    if (A_type == nullptr || B_type == nullptr ||
+        A_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ||
+        B_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ) {
+        break;
+    }
+
+    // B matrix must be constant
+    if (B_arg.Exists() && graph.GetConstantInitializer(B_arg.Name(), true) == nullptr) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/nn/matmul.cc:44]: Both inputs are required so B_arg.Exists() will always return true.

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  MatMulComputeHelper helper;
+  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b_shape, trans_a, trans_b, trans_batch_a_, trans_batch_b_));
+  Tensor* y = ctx->Output(0, helper.OutputShape());
+
+  // Bail out early if the output is going to be empty
+  if (y->Shape().Size() == 0)
+    return Status::OK();
+
+  auto* y_data = y->MutableData<float>();
+  std::unique_ptr<Tensor> packed_w_;
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/nn/matmul.cc:136]: Does anything use packed_w_?

File: onnxruntime/core/providers/cpu/math/gemm.cc (hunk 2 above)
Comment (shalvamist) [reply onnxruntime/core/providers/cpu/math/gemm.cc:172]: you are right I used the code for debugging - I will revert back the changes

File: .vscode/settings.json [hunk 10]
```diff
@@ -39,5 +39,8 @@
     "cpplint.filters": [
         "-build/include_subdir",
         "-runtime/references"
-    ]
```
Comment (fs-eire) [inline .vscode/settings.json:42]: remove changes from this file, they are not related

File: onnxruntime/core/providers/xnnpack/nn/gemm.h (hunk 5 above)
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/gemm.h]: You are right this is just a reminder from initial integration phase - I will remove it.

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc (hunk 9 above)
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/matmul.cc]: You are right - I don't use it - removed it in following commits.

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+  op0_.reset(p);
+
+  return Status::OK();
+}
+
+Status MatMul::Compute(OpKernelContext* ctx) const {
+
+  const Tensor* a = ctx->Input<Tensor>(0);
+  const auto& b_shape = b_shape_;
```
[the remark this answers is not in the document]
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/matmul.cc]: You are right, this assignment isn't needed - I removed this variable in following commits

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+  is_packed = true;
+
+  uint32_t flags = XNN_FLAG_TRANSPOSE_WEIGHTS;
+  float output_min = clip_min_max_ ? clip_min_max_->first : -INFINITY;
+  float output_max = clip_min_max_ ? clip_min_max_->second : INFINITY;
+  xnn_status status = xnn_status::xnn_status_uninitialized;
+
+  struct xnn_operator* p = nullptr;
+  this->b_shape_ = tensor.Shape();
```
[the remark this answers is not in the document]
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/matmul.cc]: OK sure, thanks for the heads up - I removed this in following commits.

File: onnxruntime/core/providers/xnnpack/detail/node_support_checker.cc (hunk 3 above)
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/detail/node_support_checker.cc]: Updated in following commits now the new ops are under xnnpack/math

File: onnxruntime/core/providers/xnnpack/nn/gemm.cc [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+  do {
+    auto input_defs = node.InputDefs();
```
[the remark this answers is not in the document]
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/gemm.cc]: OK cool - I updated the code to match the requirement

File: onnxruntime/core/providers/xnnpack/nn/gemm.cc [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+    if (!B_shape || B_shape->dim_size() > 3) {
+        break;
+    }
+
+    if (!C_shape || C_shape->dim_size() > 3) {
+        break;
+    }
+
+    // if there's a bias input it must be constant
+    if (input_defs.size() == 3) {
```
[the remark this answers is not in the document]
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/gemm.cc]: you are right I updated the code to be a bit more tidy

File: onnxruntime/core/providers/xnnpack/nn/gemm.cc [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+  return Status::OK();
+}
+
+Status Gemm::Compute(OpKernelContext* context) const {
+  //concurrency::ThreadPool* thread_pool = context->GetOperatorThreadPool();
+  const auto* A = context->Input<Tensor>(0);
+
+  // if input is empty tensor, return as nothing need to be calculated and we've set the shape for the output
+  if (M == 0 || N == 0)
+    return Status::OK();
```
[the remark this answers is not in the document]
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/gemm.cc]: thanks for catching this - I updated the code

File: onnxruntime/core/providers/xnnpack/nn/gemm.h (hunk 6 above)
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/gemm.h]: Updated in following commit - I removed the initialization

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+  info.GetAttrOrDefault<int64_t>("transB", &trans_b_attr_, 0);
+  info.GetAttrOrDefault<float>("alpha", &alpha_attr_, 1.0);
+  int64_t trans_batch_a_attr, trans_batch_b_attr;
+  info.GetAttrOrDefault<int64_t>("transBatchA", &trans_batch_a_attr, 0);
+  info.GetAttrOrDefault<int64_t>("transBatchB", &trans_batch_b_attr, 0);
+  trans_batch_a_ = trans_batch_a_attr != 0;
+  trans_batch_b_ = trans_batch_b_attr != 0;
+}
+
+Status MatMul::PrePack(const Tensor& tensor,int input_idx, AllocatorPtr alloc,
```
[the remark this answers is not in the document]
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/matmul.cc]: I called prepack since this was the convention - in pre-pack you call the set up xnnpack op - this in turn packs the B matrix into the op memory (so I guess it's kinda pre-packing). Do you have another flow in mind ? this is similar to the flow we have for Conv and Gemm.

[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/providers/xnnpack/nn/matmul.cc]: I believe you know this now, but for the benefit of anyone else reading this PR, in those cases we delayed creating the xnnpack kernel as it required the pre-packed data.

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+
+MatMul::MatMul(const OpKernelInfo& info) : OpKernel(info){
+  info.GetAttrOrDefault<int64_t>("transA", &trans_a_attr_, 0);
+  info.GetAttrOrDefault<int64_t>("transB", &trans_b_attr_, 0);
+  info.GetAttrOrDefault<float>("alpha", &alpha_attr_, 1.0);
+  int64_t trans_batch_a_attr, trans_batch_b_attr;
+  info.GetAttrOrDefault<int64_t>("transBatchA", &trans_batch_a_attr, 0);
+  info.GetAttrOrDefault<int64_t>("transBatchB", &trans_batch_b_attr, 0);
+  trans_batch_a_ = trans_batch_a_attr != 0;
+  trans_batch_b_ = trans_batch_b_attr != 0;
```
[the remark this answers is not in the document]
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/matmul.cc]: These cam from the CPU definition of MatMul op constructor. 
Thanks for highlighting this code segment, after testing the use of these attributes on several model & WASM unittest - I found they aren't needed so I removed them.

File: onnxruntime/core/providers/xnnpack/nn/matmul.cc [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+
+Status MatMul::Compute(OpKernelContext* ctx) const {
+
+  const Tensor* a = ctx->Input<Tensor>(0);
+  const auto& b_shape = b_shape_;
+  const bool trans_a = trans_a_attr_ && a->Shape().NumDimensions() != 1;
+  const bool trans_b = trans_b_attr_ && b_shape.NumDimensions() != 1;
+
+  MatMulComputeHelper helper;
+  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b_shape, trans_a, trans_b, trans_batch_a_, trans_batch_b_));
```
[the remark this answers is not in the document]
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/nn/matmul.cc]: These were used a part of the debugging process - I wanted to make sure the mat shape are aligned and received as expected. But after retesting without these parameter indeed only A & B shapes are required for correct execution of the op. I tested it on WASM unittest & several models and all passed. I removed them from the helper call.

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+      OpNodeProtoHelper info(&nc);
+
+      supported = true;
+
+    } while (false);
+
+    return supported;
+  }
+
+  Gemm::Gemm(const OpKernelInfo& info) : GemmBase(info), OpKernel(info) {
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:83]: Looks Gemm and Matmul have a lot similar code, Could we just imlemented them on the same class/file/baseclass, so we can leverage most of common code and save binary size in mobile-senario.

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "xnn_create_fully_connected_nc_f32 returned ", status);
+      }
+      op0_.reset(p);
+    }
+
+    return Status::OK();
+  }
+
+  Status Gemm::Compute(OpKernelContext* context) const {
+    // concurrency::ThreadPool* thread_pool = context->GetOperatorThreadPool();
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:163]: We have multi-thread in XNNAPCK for now, let Gemm inherit from `XnnpackKernel`, and  use `pthreadpool_t t_pool = GetThreadPool();`

Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/math/gemm.cc]: Sure thing - just pushed a commit with the threadpool integration. I tested locally and both the WASM & onnxruntime_test_all passed

File: onnxruntime/core/providers/xnnpack/math/gemm.cc (hunk 19 above)
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/math/gemm.cc]: Good idea - let me see how this can be implemented. there are a few challenges with this kind of integration but it make sense to combine the two. I will update once after the review

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+#include "core/graph/graph_utils.h"
+#include "core/framework/transpose_helper.h"
+#include "core/providers/utils.h"
+#include "core/providers/xnnpack/detail/utils.h"
+#include "core/framework/tensorprotoutils.h"
+#include "gsl/gsl-lite.hpp"
+
+#include <iostream>
+
+    namespace onnxruntime {
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:16]: Please remove the blanks ahead and justify left.

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+
+      if (input_defs.size() < 2) {
+        break;
+      }
+
+      const auto& A_arg = *input_defs[0];
+      const auto& B_arg = *input_defs[1];
+      const auto& C_arg = *input_defs[2];
+
+      // we only support float currently
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:35]: Could you also please integrate u8/s8 operators?

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+
+      if (!B_shape || B_shape->dim_size() > 3 || (B_shape->dim(0).dim_value() != 1 && B_shape->dim(1).dim_value() != 1)) {
+        break;
+      }
+
+      if (!C_shape || C_shape->dim_size() > 3) {
+        break;
+      }
+
+      ProtoHelperNodeContext nc(node_unit.GetNode());
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:73]: we can remove it if there are no attributes to be checked.

Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/math/gemm.cc:73]: seems we can check alpha/beta here

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 24]
```diff
@@ -0,0 +1,194 @@
...
+    
+    ORT_ENFORCE(info.GetAttr<float>("alpha", &alpha_).IsOK());
+    info.GetAttrOrDefault<float>("beta", &beta_, 1.f);
+
+    const auto& input_defs = node.InputDefs();
+    const NodeArg& A = *input_defs[0];
+    const NodeArg& B = *input_defs[1];
+
+    if (trans_A_ != CblasNoTrans) {
+      M = A.Shape()->dim_size() == 3 ? A.Shape()->dim(2).dim_value() : A.Shape()->dim(1).dim_value();
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:94]: `utils::GetTensorShapeFromTensorShapeProto(...)` could simper following shape retrieve operations.

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+  Status Gemm::Compute(OpKernelContext* context) const {
+    pthreadpool_t t_pool = GetThreadPool();
+    const auto* A = context->Input<Tensor>(0);
+    auto Y = context->Output(0, {M, N});
+
+    // if input is empty tensor, return as nothing need to be calculated and we've set the shape for the output
+    if (M == 0 || N == 0)
+      return Status::OK();
+    
+    xnn_status status = xnn_setup_fully_connected_nc_f32(
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:170]: we should check it also, otherwise it will be shadowed by the nest `status`

File: onnxruntime/core/providers/xnnpack/math/gemm.h [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+
+#pragma once
+
+#include "core/providers/xnnpack/xnnpack_kernel.h"
+#include "core/framework/allocator.h"
+#include "core/providers/cpu/math/gemm_base.h"
+#include "core/providers/xnnpack/detail/utils.h"
+#include "xnnpack.h"
+#include "core/common/common.h"
+#include "core/util/math.h"
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.h:12]: Could we move some headers onto `.cc` files?

Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/math/gemm.h]: I think I addressed this issue in one of my last commits - please the latest version

File: onnxruntime/core/providers/xnnpack/math/gemm.cc (hunk 25 above)
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/math/gemm.cc]: Correct - I fixed that. Thanks for highlighting this

File: onnxruntime/core/providers/xnnpack/math/matmul.cc [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+  return supported;
+}
+
+MatMul::MatMul(const OpKernelInfo& info) : XnnpackKernel(info) {
+  int64_t temp;
+  ORT_ENFORCE(info.GetAttr<int64_t>("transA", &temp).IsOK());
+  trans_A_ = temp == 0 ? CblasNoTrans : CblasTrans;
+
+  ORT_ENFORCE(info.GetAttr<int64_t>("transB", &temp).IsOK());
+  trans_B_ = temp == 0 ? CblasNoTrans : CblasTrans;
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/matmul.cc:68]: The MatMul spec doesn't have any attributes for transposing so where are these values coming from? I believe we have an internal fusion with Transpose that can add them, but that's a level 2 optimization that would happen after xnnpack sees the model. 

More concerning is that it's not clear to me how unit tests are passing given this is requiring the attributes to be provided isn't it? Would be good to validate tests are running by deliberating returning an invalid value in the results.

File: onnxruntime/core/providers/xnnpack/math/matmul.cc [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+Status MatMul::PrePack(const Tensor& tensor, int input_idx, AllocatorPtr alloc,
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/matmul.cc:71]: Why do we need prepack? It should only be used if we need to alter the layout of the weights and create a new initializer with that change.

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+
+    if (!B_shape || B_shape->dim_size() > 3) {
+      break;
+    }
+
+    if (!C_shape || C_shape->dim_size() > 3) {
+      break;
+    }
+
+    if (C_shape->dim(0).dim_value() != B_shape->dim(1).dim_value() && C_shape->dim(0).dim_value() != B_shape->dim(0).dim_value()){
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/math/gemm.cc:66]: Do we want to support attention? regarding $S(QK^T)V$, A and B are not constant initalizers.

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 30]
```diff
@@ -0,0 +1,193 @@
...
+bool Gemm::IsGemmOnnxNodeSupported(const NodeUnit& node_unit, const GraphViewer& graph) {
+  bool supported = false;
+  const onnxruntime::Node& node = node_unit.GetNode();
+
+  // use do {} while(false) so it's easier to set a breakpoint on the return
+  do {
+    ConstPointerContainer<std::vector<NodeArg*>> input_defs = node.InputDefs();
+
+    const auto alpha = node.GetAttributes().find("alpha");
+    if (!(*alpha).second.has_f() && (*alpha).second.f() != 1.0) break;
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:20]: nit: onnx spec says `alpha` must be float so if the attribute is there, has_f will always be true and you don't need to check it. same for `beta`

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 31]
```diff
@@ -0,0 +1,193 @@
...
+
+    const auto beta = node.GetAttributes().find("beta");
+    if (!(*beta).second.has_f() && (*beta).second.f() != 1.0) break;
+
+    const NodeArg* A_arg = input_defs[0];
+    const NodeArg* B_arg = input_defs[1];
+    const NodeArg* C_arg = input_defs.size() <= 2 ? nullptr : input_defs[2];
+
+    // Right now assuming C matrix is exists
+    if (!C_arg) break;
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:30]: why are we exiting if C isn't present? it's optional and a default value can be used (which is 0 as per the onnx spec) if not present. does xnnpack require it to be provided?

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 32]
```diff
@@ -0,0 +1,193 @@
...
+        A_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+      break;
+    }
+
+    // B & C matrices must be constant
+    if (!graph.IsConstantInitializer(B_arg->Name(), true)) {
+      break;
+    }
+
+    if (input_defs[2]->Exists() && !graph.IsConstantInitializer(C_arg->Name(), true)) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:45]: `C_arg->Exists()`?

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 33]
```diff
@@ -0,0 +1,193 @@
...
+
+    if (!B_shape || B_shape->dim_size() >= 3) {
+      break;
+    }
+
+    if (!C_shape || C_shape->dim_size() >= 3) {
+      break;
+    }
+    
+    if (C_shape->dim(0).dim_value() != B_shape->dim(1).dim_value() && C_shape->dim(0).dim_value() != B_shape->dim(0).dim_value()){
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:66]: Wondering if this is too strict as C only has to be broadcast-able. We use GemmBroadcastBias in the CPU implementation so some logic from there can probably be borrowed. Can add a comment in the code and do that as a separate PR if necessary.

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 34]
```diff
@@ -0,0 +1,204 @@
...
+      break;
+    }
+
+    // B & C matrices must be constant
+    if (!graph.IsConstantInitializer(B_arg->Name(), true)) {
+      break;
+    }
+
+    if (input_defs.size() == 3 && !graph.IsConstantInitializer(C_arg->Name(), true)) {
+      break;
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:51]: nit: still need to check Exists. The NodeArg might still be present with an empty name.

`if (C_arg && C_arg->Exists() && ...`

File: onnxruntime/core/providers/xnnpack/math/gemm.h [hunk 35]
```diff
@@ -0,0 +1,54 @@
...
+                 /*out*/ PrePackedWeights* prepacked_weights) override;
+
+ private:
+  const Tensor* B_{nullptr};
+
+  int64_t M_=-1;
+  int64_t K_=-1;
+  int64_t N_=-1;
+  int64_t numberOfInputs_=-1;
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.h:36]: nit: number_of_inputs_

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 36]
```diff
@@ -0,0 +1,204 @@
...
+
+  info.GetAttrOrDefault<float>("alpha", &alpha_, 1.f);
+  info.GetAttrOrDefault<float>("beta", &beta_, 1.f);
+
+  const auto& input_defs = node.InputDefs();
+  const auto* shapeA = input_defs[0]->Shape();
+  const auto* shapeB = input_defs[1]->Shape();
+
+  numberOfInputs_ = input_defs.size();
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:93]: Maybe this is the best place to check if a) the optional 3rd input is present, and if so Exists() is true or not. All other code could use this value when deciding if there's a bias input to handle

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 37]
```diff
@@ -0,0 +1,204 @@
...
+  is_packed = true;
+
+  // flags - 1 - for no transpose - 0 for transpose
+  uint32_t flags = trans_B_ == CblasTrans ? 0 : XNN_FLAG_TRANSPOSE_WEIGHTS;
+
+  float output_min = clip_min_max_ ? clip_min_max_->first : -INFINITY;
+  float output_max = clip_min_max_ ? clip_min_max_->second : INFINITY;
+
+  const float* biasData = nullptr;
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:135]: nit: bias_data

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 38]
```diff
@@ -0,0 +1,204 @@
...
+  if (numberOfInputs_ == 3) {
+    biasData = tensor.Data<float>();
+  }
+
+    xnn_status status = xnn_status::xnn_status_uninitialized;
+    struct xnn_operator* p = nullptr;
+    status = xnn_create_fully_connected_nc_f32(
+        trans_B_ == CblasNoTrans ? B_->Shape()[0] : B_->Shape()[1],  // size_t input_channels,
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:143]: nit: indentation of off.

File: onnxruntime/core/providers/xnnpack/math/gemm.h (hunk 35 above)
Comment (shalvamist) [reply onnxruntime/core/providers/xnnpack/math/gemm.h:36]: replaced the variable with a bool - indicating 3 vs. 2 inputs

File: onnxruntime/core/providers/xnnpack/math/gemm.cc [hunk 39]
```diff
@@ -0,0 +1,205 @@
...
+
+    // we only support float currently
+    const auto* A_type = A_arg->TypeAsProto();
+
+    if (A_type == nullptr || 
+        A_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+      break;
+    }
+
+    // B & C matrices must be constant
```
Comment (wejoncy) [inline onnxruntime/core/providers/xnnpack/math/gemm.cc:45]: Please align the comment with code

**Review by skottmckay — APPROVED** [review]
:shipit:
